### ISSUE TITLE
Random ContextNotActiveException with quarkus-rest-client-jackson and quarkus-grpc

### ISSUE THREAD

**Issue by alexey-plotnikoff** [quarkusio/quarkus#43372]
### Describe the bug

Example project available [here](https://github.com/alexey-plotnikoff/quarkus-bug-report1) with Quarkus version 3.14.4.
If you use [quarkus-grpc](https://quarkus.io/guides/grpc-getting-started) extension with [quarkus-rest-client-jackson](https://quarkus.io/guides/rest-client) and [quarkus-messaging-kafka](https://quarkus.io/guides/kafka) you will receive random error after 10-30 times gRPC call:
```
jakarta.enterprise.context.ContextNotActiveException: RequestScoped context was not active when trying to obtain a bean instance for a client proxy of CLASS bean [class=io.quarkus.rest.client.reactive.runtime.HeaderContainer, id=w_QyeSkhamyoJ8l04Fy3rjzMfkM]
        - you can activate the request context for a specific method using the @ActivateRequestContext interceptor binding
        at io.quarkus.arc.impl.ClientProxies.notActive(ClientProxies.java:70)
        at io.quarkus.arc.impl.ClientProxies.getSingleContextDelegate(ClientProxies.java:30)
        at io.quarkus.rest.client.reactive.runtime.HeaderContainer_ClientProxy.arc$delegate(Unknown Source)
        at io.quarkus.rest.client.reactive.runtime.HeaderContainer_ClientProxy.getHeaders(Unknown Source)
        at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientRequestFilter.filter(MicroProfileRestClientRequestFilter.java:60)
        at org.jboss.resteasy.reactive.client.spi.ResteasyReactiveClientRequestFilter.filter(ResteasyReactiveClientRequestFilter.java:21)
        at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:25)
        at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:10)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:285)
        at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:275)
        at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.method(AsyncInvokerImpl.java:215)
        at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:313)
        at am.plotnikov.bugreport1.client.GoogleClient$$QuarkusRestClientInterface.search(Unknown Source)
        at am.plotnikov.bugreport1.client.GoogleClient$$CDIWrapper.search(Unknown Source)
        at am.plotnikov.bugreport1.client.GoogleClient$$CDIWrapper_ClientProxy.search(Unknown Source)
        at am.plotnikov.bugreport1.service.KafkaService.processMessage(KafkaService.java:44)
        at am.plotnikov.bugreport1.service.KafkaService_ClientProxy.processMessage(Unknown Source)
        at am.plotnikov.bugreport1.service.KafkaService_SmallRyeMessagingInvoker_processMessage_4c7b22b897370ea933fd73d85f79da59638b74f4.invoke(Unknown Source)
        at io.smallrye.reactive.messaging.providers.AbstractMediator.lambda$invokeBlocking$15(AbstractMediator.java:191)
        at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
        at io.smallrye.mutiny.operators.uni.builders.UniCreateFromDeferredSupplier.subscribe(UniCreateFromDeferredSupplier.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
        at io.quarkus.virtual.threads.ContextPreservingExecutorService$ContextPreservingRunnable.run(ContextPreservingExecutorService.java:45)
        at java.base/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:314)
        at java.base/java.lang.VirtualThread.run(VirtualThread.java:309)
```

### Expected behavior

I should not receive random `jakarta.enterprise.context.ContextNotActiveException` error from rest client.

### Actual behavior

Random exception
`jakarta.enterprise.context.ContextNotActiveException: RequestScoped context was not active when trying to obtain a bean instance for a client proxy of CLASS bean [class=io.quarkus.rest.client.reactive.runtime.HeaderContainer, id=w_QyeSkhamyoJ8l04Fy3rjzMfkM]`

### How to Reproduce?

1. Install [grpcurl](https://github.com/fullstorydev/grpcurl)
2. Clone an [example project](https://github.com/alexey-plotnikoff/quarkus-bug-report1)
3. Run 
```
./mvnw quarkus:dev
```
4. Run
```
./start.sh
```
5. After 10-30 grpcurl call you will receive an exception from http client:
```
jakarta.enterprise.context.ContextNotActiveException: RequestScoped context was not active when trying to obtain a bean instance for a client proxy of CLASS bean [class=io.quarkus.rest.client.reactive.runtime.HeaderContainer, id=w_QyeSkhamyoJ8l04Fy3rjzMfkM]
        - you can activate the request context for a specific method using the @ActivateRequestContext interceptor binding
        at io.quarkus.arc.impl.ClientProxies.notActive(ClientProxies.java:70)
        at io.quarkus.arc.impl.ClientProxies.getSingleContextDelegate(ClientProxies.java:30)
        at io.quarkus.rest.client.reactive.runtime.HeaderContainer_ClientProxy.arc$delegate(Unknown Source)
        at io.quarkus.rest.client.reactive.runtime.HeaderContainer_ClientProxy.getHeaders(Unknown Source)
        at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientRequestFilter.filter(MicroProfileRestClientRequestFilter.java:60)
        at org.jboss.resteasy.reactive.client.spi.ResteasyReactiveClientRequestFilter.filter(ResteasyReactiveClientRequestFilter.java:21)
        at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:25)
        at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:10)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:285)
        at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:275)
        at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.method(AsyncInvokerImpl.java:215)
        at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:313)
        at am.plotnikov.bugreport1.client.GoogleClient$$QuarkusRestClientInterface.search(Unknown Source)
        at am.plotnikov.bugreport1.client.GoogleClient$$CDIWrapper.search(Unknown Source)
        at am.plotnikov.bugreport1.client.GoogleClient$$CDIWrapper_ClientProxy.search(Unknown Source)
        at am.plotnikov.bugreport1.service.KafkaService.processMessage(KafkaService.java:44)
        at am.plotnikov.bugreport1.service.KafkaService_ClientProxy.processMessage(Unknown Source)
        at am.plotnikov.bugreport1.service.KafkaService_SmallRyeMessagingInvoker_processMessage_4c7b22b897370ea933fd73d85f79da59638b74f4.invoke(Unknown Source)
        at io.smallrye.reactive.messaging.providers.AbstractMediator.lambda$invokeBlocking$15(AbstractMediator.java:191)
        at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
        at io.smallrye.mutiny.operators.uni.builders.UniCreateFromDeferredSupplier.subscribe(UniCreateFromDeferredSupplier.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
        at io.quarkus.virtual.threads.ContextPreservingExecutorService$ContextPreservingRunnable.run(ContextPreservingExecutorService.java:45)
        at java.base/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:314)
        at java.base/java.lang.VirtualThread.run(VirtualThread.java:309)

```

### Output of `uname -a` or `ver`

Darwin Amidamaru.local 23.3.0 Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

java version "21.0.2" 2024-01-16 LTS Java(TM) SE Runtime Environment Oracle GraalVM 21.0.2+13.1 (build 21.0.2+13-LTS-jvmci-23.1-b30) Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 21.0.2+13.1 (build 21.0.2+13-LTS-jvmci-23.1-b30, mixed mode, sharing)

### Quarkus version or git rev

3.14.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.8 (36645f6c9b5079805ea5009217e36f2cffd34256)

**Comment by alexey-plotnikoff** [comment]
Looks like generated http-client use RequestContext bean, but I don't have RequestContext here anymore.

**Comment by alexey-plotnikoff** [comment]
I guess problem is [here](https://github.com/quarkusio/quarkus/blob/3.14/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java#L57).

**Comment by cescoffier** [comment]
The `if` statement should check that the request scope is active. So, I'm surprised it says it is not when acquiring the bean. 

@mkouba, can you check if the if statement is the correct way to check if the request scope is active? I remember using something a bit different. 

Another thing is the usage of virtual threads with gRPC. It's a bit fragile at the moment. I wonder if we are not losing the context of something. (CC @alesj )

**Comment by geoand** [comment]
> The if statement should check that the request scope is active. So, I'm surprised it says it is not when acquiring the bean

Yeah, that looks weird

**Comment by alesj** [comment]
> Another thing is the usage of virtual threads with gRPC. It's a bit fragile at the moment. I wonder if we are not losing the context of something. (CC @alesj )

Yeah ... but we now have so many wrappers and code handling those dup-contexts - in gRPC alone, that it super hard to follow what's going on with the context ...
* https://github.com/quarkusio/quarkus/tree/main/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context
* https://github.com/quarkusio/quarkus/tree/main/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking

**Comment by alesj** [comment]
Btw, I don't see how this is affected by our gRPC code?
Since all I see is some Kafka and REST ... no gRPC :-)

**Comment by mkouba** [comment]
> @mkouba, can you check if the if statement is the correct way to check if the request scope is active? I remember using something a bit different.

@cescoffier `Arc.container().getActiveContext(RequestScoped.class) != null` should be fine, although we usually use something like `Arc.container().requestContext().isActive()`...

What could happen is that the underlying request context is invalidated after the check. This can happen when someone destroys the context state, i.e. from a different thread.

I'm not brave enough to dive into this mix of extensions (kafka, rest-client, gprc, ...) but there's a logging config to trace the request context activation: https://quarkus.io/guides/cdi-reference#how-to-enable-trace-logging-for-request-context-activation. HTH.

**Comment by cescoffier** [comment]
@alesj there is a grpc service that triggers the logic. This gRPC service uses @RunOnVirtualThread.

**Comment by alexey-plotnikoff** [comment]
@alesj gRPC is [here](https://github.com/alexey-plotnikoff/quarkus-bug-report1/blob/main/src/main/java/am/plotnikov/bugreport1/service/HelloWorldService.java#L12).
Everything's fine if I use REST only without [quarkus-grpc](https://github.com/alexey-plotnikoff/quarkus-bug-report1/blob/main/pom.xml#L47) library and virtual threads.

**Comment by alexey-plotnikoff** [comment]
@cescoffier look at this. @RunOnVirtualThread is disabled for gRPC, but enabled for Kafka Incoming:

<img width="1391" alt="Screenshot 2024-09-20 at 09 27 57" src="https://github.com/user-attachments/assets/9963219c-9042-48c3-88b1-e822a934da15">

**Comment by alexey-plotnikoff** [comment]
And both @RunOnVirtualThread are disabled. Hmmm:

<img width="1374" alt="Screenshot 2024-09-20 at 09 33 56" src="https://github.com/user-attachments/assets/85e6d59b-3b36-4141-9afa-db38dec767d6">

**Comment by cescoffier** [comment]
Thanks!  At least it has eliminated my first guess. 

Second guess: we need to check is the `sayHello` method is invoked on a duplicated context and has the request scoped enabled. It should be done automatically, but let's make sure of it. 

@alesj can you verify this second guess? 

@alexey-plotnikoff, let's try something to see if it improves. Can you annotate `sayHello` with `@ActivateRequestScope`.

**Comment by alexey-plotnikoff** [comment]
@cescoffier I did it. And looks like it didn't help :(

<img width="1389" alt="Screenshot 2024-09-20 at 14 51 45" src="https://github.com/user-attachments/assets/9ee4a9e1-6f6f-4176-bbff-68fabb0b462a">

**Comment by alexey-plotnikoff** [comment]
@cescoffier why rest-client uses current request context? It is not clear for me.

My service should receive request from external system, run http-client in separate thread to release external system as soon as possible.

Temporary I've been replaced rest-client with [HttpClient](https://openjdk.org/groups/net/httpclient/intro.html) and looks like it works now.

**Comment by geoand** [comment]
> @cescoffier why rest-client uses current request context? It is not clear for me.

So it can propagate incoming headers if necessary

**Comment by alexey-plotnikoff** [comment]
Btw, the REST Client internally uses [the Vert.x HTTP Client](https://vertx.io/docs/apidocs/io/vertx/core/http/HttpClient.html) that executes HTTP in own pool, but with virtual threads I guess it now actual anymore.

**Comment by alexey-plotnikoff** [comment]
> So it can propagate incoming headers if necessary

@geoand For what? Opentelemetry?

**Comment by geoand** [comment]
Among other

**Comment by alexey-plotnikoff** [comment]
I want to share some additional data. Maybe it can help.
If I replace incoming annotation with something like this:
```
...
var factory = Thread.ofVirtual().name("my-virtual-").factory();
var executorService = Executors.newThreadPerTaskExecutor(factory);
...
executorService.submit(() -> restHttpClient.post("something"));
...
```
I see this (native build, with pod restarts).

![IMG_9264 Medium](https://github.com/user-attachments/assets/e66eb08a-c041-4e29-9fc5-fb4e3c2835ea)

Looks like a memory leak. I guess, maybe context duplicates to wrong place and not clears.

**Comment by alesj** [comment]
Aleš Justin: Let me add more log to that Kafka service

Aleš Justin:

gRPC - Vertx.context=[{java.lang.Object@31c6a4a4=true, GRPC_CONTEXT=io.grpc.Context@2fe528d7, io.quarkus.vertx.cdi-current-contextjakarta.enterprise.context.RequestScoped=io.quarkus.arc.impl.RequestContext$RequestContextState@702694bf}]

gRPC - isDup=true

gRPC - isActive=true

2024-09-20 11:06:06,924 INFO  [am.plo.bug.ser.HelloWorldService] (quarkus-virtual-thread-297) ----

2024-09-20 11:06:06,924 INFO  [am.plo.bug.ser.HelloWorldService] (quarkus-virtual-thread-297) gRPC finished

2024-09-20 11:06:06,924 INFO  [am.plo.bug.ser.KafkaService] (quarkus-virtual-thread-298) HTTP Client Request started

Kafka - Vertx.context=[{java.lang.Object@31c6a4a4=true, GRPC_CONTEXT=io.grpc.Context@2fe528d7, io.quarkus.vertx.cdi-current-contextjakarta.enterprise.context.RequestScoped=io.quarkus.arc.impl.RequestContext$RequestContextState@702694bf}]

Kafka - isDup=true

Kafka - isActive=false

Aleš Justin: ... hmmm ... isActive in Kafka == false


@Incoming("in-memory-google")
@RunOnVirtualThread
public void processMessage(String text) {
    try {
        log.info("HTTP Client Request started");
        System.out.println("Kafka - Vertx.context=" + Vertx.currentContext());
        System.out.println("Kafka - isDup=" + VertxContext.isOnDuplicatedContext());
        System.out.println("Kafka - isActive=" + Arc.container().requestContext().isActive());
        log.info("----");
// clientMap.get("GOOGLE").search(text);
client.search(text);

**Comment by alexey-plotnikoff** [comment]
@alesj Could you format your message above? 
Basic writing and formatting syntax on GitHub is available [here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax).
Thank you! :)

**Comment by alesj** [comment]
Just dumping the info asap :)
Will do.

**Comment by alesj** [comment]
If I comment out RunOnVT + add Blocking
```
    @Incoming("in-memory-google")
    //@RunOnVirtualThread
    @ActivateRequestContext
    @Blocking
    public void processMessage(String text) {
        try {
            log.info("HTTP Client Request started");
            System.out.println("Kafka - Vertx.context=" + Vertx.currentContext());
```

I get this:

```
2024-09-27 11:35:15,771 INFO  [am.plo.bug.ser.KafkaService] (vert.x-worker-thread-1) HTTP Client Request started
Kafka - Vertx.context=[{java.lang.Object@75938fd2=true, io.quarkus.vertx.cdi-current-contextjakarta.enterprise.context.RequestScoped=io.quarkus.arc.impl.RequestContext$RequestContextState@2bc1b075}]
Kafka - isDup=true
Kafka - isActive=true
2024-09-27 11:35:15,772 INFO  [am.plo.bug.ser.KafkaService] (vert.x-worker-thread-1) ----
```

So something is not activating the context with that @RunOnVT ... @ozangunalp ... ?

**Comment by cescoffier** [comment]
@alesj  / @ozangunalp Any progress? Ozan is doing some work around context propagation, but it's rather in the opposite direction (sometimes the request scope gets propagated while we don't want that)

**Comment by alesj** [comment]
I can re-check if something changed, for the better.
Did we get a combination of annotations that made this somehow work?
(afk atm … AI workshop)

**Comment by alexey-plotnikoff** [comment]
I checked some weeks ago on version 3.16.2. Problem was the same.

**Comment by ozangunalp** [comment]
I'll check the reproducer from this issue if my coming changes fixes it correctly.

**Comment by gregopet** [comment]
We hit the same bug without gRPC - we trigger a REST client call on a virtual thread (via `@VirtualThreads ExecutorService` since we want the regular request flow to continue independently), and apparently someone also annotated the new thread's `run` with `@ActivateRequestContext`, but it doesn't help.

**Comment by alesj** [comment]
@gregopet can you provide a simple reproducer app?

**Comment by ozangunalp** [comment]
@gregopet I understand that you want 
> the regular request flow to continue independently

but the RequestScoped context is bound to the REST request you receive. When this one returns, it'll terminate and destroy the request context. 

I reckon you've it sporadically because in some cases, the processing you do inside VT terminates before your request returns.

**Comment by gregopet** [comment]
I put together a repro case in https://github.com/gregopet/quarkus-bug-report

It didn't start manifesting until I created the REST client using RestClientBuilder (which is what we do in the original app as well).

**Comment by alesj** [comment]
> but the RequestScoped context is bound to the REST request you receive. When this one returns, it'll terminate and destroy the request context.

@gregopet this ^^ is not the case with your issue?

**Comment by gregopet** [comment]
@alesj I don't think so. If you look at `MicroProfileRestClientRequestFilter` there's an if clause there

```if (Arc.container().getActiveContext(RequestScoped.class) != null) {```

Whenever that evaluates to true, that exception is thrown - but this only happens sporadically (e.g. 5 times out of 100, when running test requests in a loop). Most of the time the request is not there, `getActiveContext` returns `null` and code continues without exceptions.

**Comment by kpagratis** [comment]
We're seeing the same behavior with quarkus rest.

**Comment by alexey-plotnikoff** [comment]
I use [HttpClient](https://docs.oracle.com/en/java/javase/21/docs/api/java.net.http/java/net/http/HttpClient.html) from java.net.http package and problem is gone.
I think the problem is in quarkus rest-client extension.

**Comment by geoand** [comment]
I checked the reproducer mentioned [here](https://github.com/quarkusio/quarkus/issues/43372#issuecomment-2640523320) and it indeed reproduces the issue faithfully (and only depends on the REST Client).

However, when using taking

```java
    @VirtualThreads
    ExecutorService vThreads;
```

out of the equasion, the issue goes away, so my guess is that `ContextPreservingExecutorService` doesn't work as expected.

**Comment by geoand** [comment]
You are right!

I didn't actually pay enough attention to either what the application is doing or the details of the exception.

So indeed in the case of the sample, a task is submitted in fire and forget mode and depending on timing, when the client attempts to pull the incoming headers, it fails

**Comment by alesj** [comment]
> [@geoand](https://github.com/geoand) and also, the issue seems to disappear if the client isn't created through code using `RestClientBuilder`

Which is weird, since what @ozangunalp says, makes sense ... dunno how this - non RestClientBuilder - could "fix" that?!

**Comment by geoand** [comment]
It's probably just a matter of timming, i.e. using the builder vs the CDI client takes longer to create and thus increases the chances of the original HTTP request having completed

**Comment by gregopet** [comment]
Hmm no, I'm sorry, you're right, Aleš: who knows what I tried before completing the prototype, but if I go back and inject the service with `@RestClient` I can also reproduce the issue - sorry for the misleading info!

**Comment by geoand** [comment]
No problem :)

**Comment by geoand** [comment]
@kpagratis do you have perhaps have a sample application that we can use to reproduce the problem?

**Comment by geoand** [comment]
That would be great, thanks!

**Comment by linxux** [comment]
Version: Quarkus 3.19.2
After reactivating the browser tab, the first request frequently reports an error - "RequestScoped context was not active".

Some depdencies,
```
    implementation 'io.quarkus:quarkus-rest'
    implementation 'io.quarkus:quarkus-rest-jackson'
    implementation("io.quarkus:quarkus-smallrye-context-propagation")

    implementation 'io.quarkus:quarkus-oidc'
    implementation 'io.quarkus:quarkus-arc'
```

```
13:09:34 ERROR [io.bi.co.ht.ex.ThrowableMapper]] (vert.x-eventloop-thread-1) exception: jakarta.enterprise.context.ContextNotActiveException: RequestScoped context was not active when trying to obtain a bean instance for a client proxy of CLASS bean [class=io.quarkus.vertx.http.runtime.CurrentVertxRequest, id=0_6n6EmChCiiDdd8HelptG_A0AE]
	- you can activate the request context for a specific method using the @ActivateRequestContext interceptor binding
	at io.quarkus.arc.impl.ClientProxies.notActive(ClientProxies.java:76)
	at io.quarkus.arc.impl.ClientProxies.getSingleContextDelegate(ClientProxies.java:32)
	at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ClientProxy.arc$delegate(Unknown Source)
	at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ClientProxy.setCurrent(Unknown Source)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusCurrentRequest.set(QuarkusCurrentRequest.java:33)
	at org.jboss.resteasy.reactive.server.core.CurrentRequestManager.set(CurrentRequestManager.java:12)
	at org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext.handleRequestScopeActivation(ResteasyReactiveRequestContext.java:656)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.handleRequestScopeActivation(QuarkusResteasyReactiveRequestContext.java:39)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.requireCDIRequestScope(AbstractResteasyReactiveContext.java:264)
	at org.jboss.resteasy.reactive.server.handlers.ResourceRequestFilterHandler.handle(ResourceRequestFilterHandler.java:46)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:131)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at org.jboss.resteasy.reactive.server.vertx.VertxResteasyReactiveRequestContext$1$1.handle(VertxResteasyReactiveRequestContext.java:84)
	at org.jboss.resteasy.reactive.server.vertx.VertxResteasyReactiveRequestContext$1$1.handle(VertxResteasyReactiveRequestContext.java:81)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)

```

**Comment by geoand** [comment]
@linxux do you have a simple reproducer for that? If so, please attach it and we can take a look.

**Comment by ozangunalp** [comment]
If I remember correctly the initial reproducer used an internal messaging channel. So it is fixed now with #46174

**Comment by geoand** [comment]
Thanks @ozangunalp. Let's close this then an there are problems with REST Client only (no gRPC) and someone has a reproducer, please open a new issue

**Comment by alexey-plotnikoff** [comment]
Wait a minute, I can check this. My [example project](https://github.com/alexey-plotnikoff/quarkus-bug-report1) is still available.

**Comment by gregopet** [comment]
@geoand I tried my test case with Quarkus 3.21.1 and it still fails - should this version already contain the above mentioned fix?

**Comment by geoand** [comment]
> Thanks [@ozangunalp](https://github.com/ozangunalp). Let's close this then an there are problems with REST Client only (no gRPC) and someone has a reproducer, please open a new issue

@gregopet see ^
